### PR TITLE
Improved performances for isPrime

### PR DESCRIPTION
**Pull request by arguiot**
The prime test performance has been improved compared to the previous version.

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks Arthur! Does the algorithm you implemented have a specific name?

Can you please have a look at the failing tests "`n` is undefined? I also see a semicolon, that will let the linter break (`npm run lint`).

**Comment by arguiot** [comment]
Alright, I will try to fix that as soon as possible. I don’t know if it has a specific name, I use it in [TheoremJS](https://theorem.js.org) and I compared it with yours, and mine seems to be faster. It’s based on the multiple of 6 approach, described [here](http://mathforum.org/library/drmath/view/56068.html).

File: src/function/utils/isPrime.js [hunk 1]
```diff
@@ -35,39 +35,37 @@ export const createIsPrime = /* #__PURE__ */ factory(name, dependencies, ({ type
...
-      if (x.lt(2)) {
-        return false
-      }
-      if (x.equals(2)) {
-        return true
-      }
-      if (x.mod(2).isZero()) {
-        return false
+    BigNumber: function (n) {
+      if (n.toNumber() * 0 !== 0) {
```
Comment (Yaffle) [inline src/function/utils/isPrime.js]: PeterOlson's lib has `isPrime` for BigIntegers - https://github.com/peterolson/BigInteger.js/blob/master/BigInteger.js#L917 
I believe, it is based on https://en.wikipedia.org/wiki/Miller–Rabin_primality_test#Deterministic_variants .

Comment (arguiot) [reply src/function/utils/isPrime.js]: Yeah, but if you read the Wikipedia article, you discover that's it's not 100% accurate. A such algorithm is definitely fast, but as we can't be certain if a given n is a prime or not, I would strongly suggest not to use it.

**Comment by josdejong** [comment]
@arguiot how can I easily test that your implementation is faster than the current one?

**Comment by arguiot** [comment]
@josdejong You can use the [benchmark module](https://www.npmjs.com/package/benchmark), and test it. Here is how I would do it:
```js
const { Suite } = require('benchmark');
const primeval = function mathjsPrime() {} // please copy your function here
const arguiot = function arguiotPrime() {} // please copy my function here

const primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199];
const notprimes = [0, 1, 4, 6, 8, 10, 14, 15, 25, 30, 77, 100, 625, 932];
const carmichaels = [561, 1105, 1729, 2465, 2821, 6601, 8911, 41041, 825265, 321197185]; // more challenging numbers

function bench(name) {
	console.log(`\n# ${name}`);
	return new Suite().on('cycle', x => console.log(x.target.toString()));
}


bench('Primes')
	.add('arguiot', () => {
		primes.forEach(arguiot);
	})
	.add('MathJS', () => {
		primes.forEach(primeval);
	})
	.run();


bench('Not Primes')
	.add('arguiot', () => {
		notprimes.forEach(arguiot);
	})
	.add('MathJS', () => {
		notprimes.forEach(primeval);
	})
	.run();

bench('Carmichael Numbers')
	.add('arguiot', () => {
		carmichaels.forEach(arguiot);
	})
	.add('MathJS', () => {
		carmichaels.forEach(primeval);
	})
	.run();
```

**Comment by ericman314** [comment]
Ooh, if we're running benchmarks, why not go big?

**Primes:**
524287
2147483647
87178291199
4398042316799
99194853094755497
2305843009213693951
11111111111111111111111

**Not primes:**
2199023255551
8796093022207
140737488355327
9007199254740991 (MAX_SAFE_INTEGER)
57646075230343487
147573952589676412927

**Comment by arguiot** [comment]
Why not, but we'll have to describe these numbers as BigInteger then.

**Comment by josdejong** [comment]
I've added a small benchmark: [test/benchmark/prime.js](https://github.com/josdejong/mathjs/blob/develop/test/benchmark/prime.js)

On my machine this gives the following results:

```
before improved algorithm of Arthur:

primes x 105 ops/sec ±0.41% (76 runs sampled)
not primes x 15,850 ops/sec ±0.30% (92 runs sampled)
carmichaels x 10,791,643 ops/sec ±0.23% (95 runs sampled)


after improved algorithm of Arthur:

primes x 158 ops/sec ±0.73% (80 runs sampled)
not primes x 23,573 ops/sec ±0.56% (94 runs sampled)
carmichaels x 10,595,211 ops/sec ±0.51% (96 runs sampled)
```

So that's a serious improvement :+1: . Thanks Arthur!

File: src/function/utils/isPrime.js (hunk 1 above)
Comment (Yaffle) [reply src/function/utils/isPrime.js:56]: @arguiot , are you sure? I could be wrong, but I think, what Wikipedia article says is that the algorithm is based on something unproved, but it is 100% accurate

**Comment by josdejong** [comment]
Thanks for sharing @vitaly-t . Feel free to create a PR improving `isPrime` further 😄

**Comment by Yaffle** [comment]
the Miller-Rabin test is even faster if to implement it for small integers and if to run only after checking divisors <= 1000

**Comment by vitaly-t** [comment]
> the Miller-Rabin test is even faster if to implement it for small integers

I'm not sure about small numbers... [my implementation](https://github.com/vitaly-t/primes-generator/blob/main/src/is-prime.ts) can check the first million primes in just 217ms ([benchmarks](https://github.com/vitaly-t/primes-generator/tree/main/test)). How much faster Miller-Rabin can do it, though? From my tests, Miller-Rabin is good for larger numbers, while for small ones it is an overkill.

**Comment by arguiot** [comment]
I already implemented a Miller Rabin testing function in another language, but the code is pretty easy to read and reproduce in JS I think, so it's definitely translatable for `Number`.

I leave the source code here, I don't have the time right now to work on it:
https://github.com/arguiot/Euler/blob/master//Sources/Euler/NumberTheory/isPrime.swift#L88-L127

**Pull request merged**